Update Devfile Import to use Kubernetes YAML definitions

maysunfaisal opened the pull request:
### Description
This PR updates the Devfile handler backend code to consume the Kubernetes YAML file definitions. For now, the devfile samples only have the `Deployment` definition in the YAML file but the next course of action would be to update the `Service`, `Route` definitions in the devfile sample

Fixes https://github.com/devfile/api/issues/924

This PR just added the ability to read from Kubernetes YAML if the Devfile Kubernetes Component has `inlined` definition. If inlined is not present, it resorts to the previous way of generating Kubernetes resources. This should be compatible with previous OpenShift versions and the master branch

This PR also refactored some of the devfile backend code to keep it more clean and organized. Added tests to the devfile pkg since ODC didnt have go tests. Coverage %:

<img width="1331" alt="Screen Shot 2022-09-22 at 1 33 59 PM" src="https://user-images.githubusercontent.com/31771087/191813939-002f743c-ac5b-4e89-9837-21c4e712d27d.png">


### How to test
(Check history for prev edit)
- Devfile Import should work fine from the UI
- If you are interested in testing the new approach, go to devfile-handler.go and update the following lines. Build and test it out by only importing a _python devfile sample_
```
         // Get devfile content and parse it using a library call in the future
	// devfileContentBytes := []byte(data.Devfile.DevfileContent)
	convert := true
	devfileObj, _, err = devfile.ParseDevfileAndValidate(parser.ParserArgs{
		// Data: devfileContentBytes,
		URL: "https://raw.githubusercontent.com/devfile-samples/devfile-sample-python-basic/87e269d5e018c904ba0a72b53292b3a9ed0b8b37/devfile.yaml",
		ConvertKubernetesContentInUri: &convert,
	})
```

jerolimov commented on pkg/server/devfile-handler.go:
```diff
@@ -61,10 +62,15 @@ func (s *Server) devfileHandler(w http.ResponseWriter, r *http.Request) {
+	convertUriToInlined := true
+	devfileObj, _, err = devfile.ParseDevfileAndValidate(parser.ParserArgs{
+		// Data:                          devfileContentBytes,
+		URL:                           "https://raw.githubusercontent.com/devfile-samples/devfile-sample-python-basic/main/devfile.yaml",
```
I understand that this is WIP, but as a reminder that we don't want merge this URL unexpectedly.

maysunfaisal replied:
removed

maysunfaisal commented:
@yangcao77 could you PTAL from Devfile perspective?

maysunfaisal commented on pkg/server/devfile-handler.go:
```diff
@@ -23,6 +23,7 @@ import (
 	"github.com/devfile/library/pkg/devfile/generator"
 	"github.com/devfile/library/pkg/devfile/parser"
 	"github.com/devfile/library/pkg/devfile/parser/data/v2/common"
+	"github.com/devfile/library/pkg/testingutil/filesystem"
```
updated

yangcao77 commented on pkg/devfile/resources.go:
```diff
@@ -0,0 +1,258 @@
+				}
+
+				// Service
+				serviceResource, err := GetService(devfileObj, imageComponentFilter, dockerImagePort)
```
why we want to pass in the imageComponentFilter as an arg? can this be defined within `GetService`?

yangcao77 replied:
why want to pass in image component filter as an arg? can this be defined within the function itself?

yangcao77 commented on pkg/devfile/resources.go:
```diff
@@ -0,0 +1,258 @@
+		if _, ok := deployAssociatedComponents[component.Name]; ok && component.Kubernetes != nil {
+			var src parser.YamlSrc
+
+			if component.Kubernetes.Inlined != "" {
```
instead of a if-else statement. can check `if component.Kubernetes.Inlined == ""`, and do the dev preview way. since it's going to break out if falls in that if statement. the `src` definition can be defined with all the stuffs starting at [line 186](https://github.com/openshift/console/blob/f1d0fd8d641b0dbb6810681be25de717624f7ca7/pkg/devfile/resources.go#L186)

yangcao77 commented on pkg/devfile/resources.go:
```diff
@@ -0,0 +1,258 @@
+			}
+
+			for _, endpoint := range component.Kubernetes.Endpoints {
+				if endpoint.Exposure != devfilev1.NoneEndpointExposure {
```
should not use an `internal` endpoint to create route neither

yangcao77 commented on pkg/devfile/resources.go:
```diff
@@ -0,0 +1,258 @@
+			for _, endpoint := range component.Kubernetes.Endpoints {
+				if endpoint.Exposure != devfilev1.NoneEndpointExposure {
+					var isSecure bool
+					if endpoint.Secure != nil {
```
you don't need to check for this. it's been done in parser already https://github.com/kim-tsao/library/blob/4c7c5dc73eee13b3db599a7e5fd2b266420cd390/pkg/devfile/parser/parse.go#L702

maysunfaisal replied:
I would still need to check if its nil because I am accessing it like *endpoint.Secure; I didnt check initially and it panicked in one of my tests with an empty `secure`

yangcao77 replied:
in unit test, yes, you can assign a nil to `endpoint.Secure`. but in actual case, all boolean values will be set before returning `devfileObj`

yangcao77 commented on pkg/devfile/resources.go:
```diff
@@ -0,0 +1,258 @@
+		if len(errMsg) != 0 {
+			errMsg = errMsg + ", "
+		}
+		errMsg = errMsg + "no service definition was found in the devfile sample"
```
instead of doing the concatenating. can use `github.com/hashicorp/go-multierror`?
and 
```
errMsg = multierror.Append(errMsg, fmt.Errorf("no service definition was found in the devfile sample"))
```

maysunfaisal replied:
oh should have used this

yangcao77 reviewed (commented):
generally looks good from devfile perspective. a couple suggestions added

maysunfaisal commented on pkg/devfile/resources.go:
```diff
@@ -0,0 +1,258 @@
+
+				dockerImagePort := devfileObj.Data.GetMetadata().Attributes.GetString("alpha.dockerimage-port", &err)
+				if err != nil {
+					errMsg := fmt.Sprintf("Failed to get the Dockerfile location from devfile metadata attribute 'alpha.build-dockerfile': %v", err)
```
it was already in master branch, I didnt introduce this; I just merely moved the code. But let me see what i can do

kadel commented on pkg/devfile/resources.go:
```diff
@@ -0,0 +1,237 @@
+}
+
+// GetDeployResource gets the deployment resource
+func GetDeployResource(name string, devfileObj parser.DevfileObj, filterOptions common.DevfileOptions) (appsv1.Deployment, error) {
```
Why we can't use what is already in there in `devfile/library`?  (https://github.com/devfile/library/blob/a7dbafc0c2807db47d9ee7fa2a1cf19998b0236f/pkg/devfile/generator/generators.go#L180)

jerolimov commented on pkg/devfile/types.go:
```diff
@@ -1,4 +1,4 @@
-package server
+package devfile
```
@jhadvig What's you opinion on this move? We have also `helm`, `knative`, `terminal` on this level, so it might be ok?

Alternatively, we could move all these packages into a `devconsole` package (another time)? Wdyt? :man_shrugging:

jerolimov reviewed (commented):
Just found some nits, this should not block us from pushing this forward.

jerolimov reviewed (changes requested):
Tested the following cases:

1. Imported all devfile samples on 4.12 (Kubernetes resources are ignored)
    * Basic Quarkus example doesn't start successfully, this issue is addressed here: https://github.com/devfile-samples/devfile-sample-code-with-quarkus/pull/7
2. Imported all devfile samples with this PR
    * All samples could be imported successful :+1:
    * Basic Quarkus example could be imported from https://github.com/yangcao77/devfile-sample-code-with-quarkus
    * Compared the `/api/devfile/` response and it is identical to the old version [2] :+1: 
3. Added the following code to `packages/dev-console/src/components/import/devfile/devfileHooks.ts` [1] to insert some `inlined` kubernetes resources from https://github.com/devfile-samples/devfile-sample-dotnet60-basic
    * The `/api/devfile/` response contains an updated `Deployment` and `Service` and these look fine to me :+1:
    * :warning: Unfortunately, the import doesn't work because the `Route` resource is broken. You can see the full JSON response below. [3] :warning:

In the console the dry-run fail already:

![image](https://user-images.githubusercontent.com/139310/193544266-0d6c9e38-800e-4468-bbe1-8bbb7d3367f7.png)

The `Deployment` diff looks good to me:

![image](https://user-images.githubusercontent.com/139310/193545012-7cf2a847-ca23-4800-9ec1-eaf77124be1b.png)

The `Service` diff looks also fine:

![image](https://user-images.githubusercontent.com/139310/193545113-6df085c3-ef5f-4c5b-8dba-5d743e7bf6cd.png)

Here you see the broken `Route`:

![image](https://user-images.githubusercontent.com/139310/193545350-31c8fa38-f307-4245-880e-c0ca3c6761ca.png)


[1] code snippet for `devfileHooks.ts` to test inlined yaml

```
tsx
import { safeLoad, safeDump } from 'js-yaml';
```

```tsx
    /* eslint-disable no-console */
    console.log('xxx oldDevfileContent', devfileContent);

    const parsedDevfileContent = safeLoad(devfileContent);
    console.log('xxx parsed devfileContent', devfileContent);

    if (
      parsedDevfileContent.metadata.name === 'dotnet60' &&
      Array.isArray(parsedDevfileContent.components)
    ) {
      console.log('xxx update devfile with metadata.name', parsedDevfileContent.metadata.name);
      parsedDevfileContent.components.forEach((component) => {
        if (component.kubernetes && component.kubernetes.uri === 'kubernetes/deployment.yaml') {
          component.kubernetes = {
            inlined: `kind: Deployment
apiVersion: apps/v1
metadata:
  name: devfile-dotnet-deploy
spec:
  replicas: 1
  selector:
    matchLabels:
      app: devfile-dotnet-deploy
  template:
    metadata:
      labels:
        app: devfile-dotnet-deploy
    spec:
      containers:
        - name: main
          image: "{{CONTAINER_IMAGE}}"
          imagePullPolicy: Always
          resources: {}
`,
          };
        }
        if (component.kubernetes && component.kubernetes.uri === 'kubernetes/service.yaml') {
          component.kubernetes = {
            inlined: `apiVersion: v1
kind: Service
metadata:
  labels:
    app: dotnet-deploy
  name: devfile-dotnet-deploy
spec:
  ports:
    - name: http-8081
      port: 8081
      protocol: TCP
      targetPort: 8081
  selector:
    app: devfile-dotnet-deploy
  type: LoadBalancer
`,
          };
        }
      });
    } else {
      console.log('xxx skip devfile with metadata.name', parsedDevfileContent.metadata.name);
    }

    const newDevfileContent = safeDump(parsedDevfileContent);
    console.log('xxx newDevfileContent', newDevfileContent);

    return {
      name,
      git: { URL: url, ref, dir: prefixDotSlash(dir) },
      devfile: { devfileContent: newDevfileContent, devfilePath: `${smartSlashDir}${devfilePath}` },
    };
  }, [name, url, devfileContent, ref, dir, smartSlashDir, devfilePath]);
```


[2] old and new json response from `/api/devfile/` (when no inlined kubernetes resource is used)

```json
{
  "imageStream": {
    "kind": "ImageStream",
    "apiVersion": "image.openshift.io/v1",
    "metadata": {
      "creationTimestamp": null
    },
    "spec": {
      "lookupPolicy": {
        "local": false
      }
    },
    "status": {
      "dockerImageRepository": ""
    }
  },
  "buildResource": {
    "kind": "BuildConfig",
    "apiVersion": "build.openshift.io/v1",
    "metadata": {
      "creationTimestamp": null
    },
    "spec": {
      "source": {
        "type": "Git",
        "git": {
          "uri": "https://github.com/devfile-samples/devfile-sample-dotnet60-basic.git"
        },
        "contextDir": "."
      },
      "strategy": {
        "type": "Docker",
        "dockerStrategy": {
          "dockerfilePath": "docker/Dockerfile"
        }
      },
      "output": {
        "to": {
          "kind": "ImageStreamTag",
          "name": "dotnet60-basic:latest:latest"
        }
      },
      "resources": {},
      "postCommit": {},
      "nodeSelector": null
    },
    "status": {
      "lastVersion": 0
    }
  },
  "deployResource": {
    "kind": "Deployment",
    "apiVersion": "apps/v1",
    "metadata": {
      "creationTimestamp": null
    },
    "spec": {
      "selector": {
        "matchLabels": {
          "app": "dotnet60-basic"
        }
      },
      "template": {
        "metadata": {
          "creationTimestamp": null
        },
        "spec": {
          "containers": [
            {
              "name": "dotnet",
              "image": "registry.access.redhat.com/ubi8/dotnet-60:6.0",
              "command": [
                "tail",
                "-f",
                "/dev/null"
              ],
              "ports": [
                {
                  "name": "http-8080",
                  "containerPort": 8080,
                  "protocol": "TCP"
                }
              ],
              "env": [
                {
                  "name": "CONFIGURATION",
                  "value": "Debug"
                },
                {
                  "name": "STARTUP_PROJECT",
                  "value": "app.csproj"
                },
                {
                  "name": "ASPNETCORE_ENVIRONMENT",
                  "value": "Development"
                },
                {
                  "name": "PROJECTS_ROOT",
                  "value": "/projects"
                },
                {
                  "name": "PROJECT_SOURCE",
                  "value": "/projects"
                }
              ],
              "resources": {},
              "imagePullPolicy": "Always"
            }
          ]
        }
      },
      "strategy": {
        "type": "Recreate"
      }
    },
    "status": {}
  },
  "service": {
    "kind": "Service",
    "apiVersion": "v1",
    "metadata": {
      "creationTimestamp": null
    },
    "spec": {
      "ports": [
        {
          "name": "http-8080",
          "port": 8080,
          "targetPort": 8080
        },
        {
          "name": "http-8081",
          "port": 8081,
          "targetPort": "8081"
        }
      ]
    },
    "status": {
      "loadBalancer": {}
    }
  },
  "route": {
    "kind": "Route",
    "apiVersion": "route.openshift.io/v1",
    "metadata": {
      "creationTimestamp": null
    },
    "spec": {
      "path": "/",
      "to": {
        "kind": "Service",
        "name": "dotnet60-basic",
        "weight": null
      },
      "port": {
        "targetPort": "8081"
      }
    },
    "status": {}
  }
}
```


[3] updated `/api/devfile/` json response with inlined YAML

```json
{
  "imageStream": {
    "kind": "ImageStream",
    "apiVersion": "image.openshift.io/v1",
    "metadata": {
      "creationTimestamp": null
    },
    "spec": {
      "lookupPolicy": {
        "local": false
      }
    },
    "status": {
      "dockerImageRepository": ""
    }
  },
  "buildResource": {
    "kind": "BuildConfig",
    "apiVersion": "build.openshift.io/v1",
    "metadata": {
      "creationTimestamp": null
    },
    "spec": {
      "source": {
        "type": "Git",
        "git": {
          "uri": "https://github.com/devfile-samples/devfile-sample-dotnet60-basic.git"
        },
        "contextDir": "."
      },
      "strategy": {
        "type": "Docker",
        "dockerStrategy": {
          "dockerfilePath": "docker/Dockerfile"
        }
      },
      "output": {
        "to": {
          "kind": "ImageStreamTag",
          "name": "dotnet60-basic:latest:latest"
        }
      },
      "resources": {},
      "postCommit": {},
      "nodeSelector": null
    },
    "status": {
      "lastVersion": 0
    }
  },
  "deployResource": {
    "kind": "Deployment",
    "apiVersion": "apps/v1",
    "metadata": {
      "name": "devfile-dotnet-deploy",
      "creationTimestamp": null
    },
    "spec": {
      "replicas": 1,
      "selector": {
        "matchLabels": {
          "app": "devfile-dotnet-deploy"
        }
      },
      "template": {
        "metadata": {
          "creationTimestamp": null,
          "labels": {
            "app": "devfile-dotnet-deploy"
          }
        },
        "spec": {
          "containers": [
            {
              "name": "main",
              "image": "{{CONTAINER_IMAGE}}",
              "resources": {},
              "imagePullPolicy": "Always"
            }
          ]
        }
      },
      "strategy": {}
    },
    "status": {}
  },
  "service": {
    "kind": "Service",
    "apiVersion": "v1",
    "metadata": {
      "name": "devfile-dotnet-deploy",
      "creationTimestamp": null,
      "labels": {
        "app": "dotnet-deploy"
      }
    },
    "spec": {
      "ports": [
        {
          "name": "http-8081",
          "protocol": "TCP",
          "port": 8081,
          "targetPort": 8081
        }
      ],
      "selector": {
        "app": "devfile-dotnet-deploy"
      },
      "type": "LoadBalancer"
    },
    "status": {
      "loadBalancer": {}
    }
  },
  "route": {
    "metadata": {
      "creationTimestamp": null
    },
    "spec": {
      "to": {
        "kind": "",
        "name": "",
        "weight": null
      }
    },
    "status": {}
  }
}
```

maysunfaisal commented:
@jerolimov Thanks for the detailed testing and reports. The routes need to be defined in the devfile this way https://github.com/devfile-samples/devfile-sample-code-with-quarkus/pull/7/files#diff-23d39e3ae288b080d8aa0995928cd3c432f887a75f0ab3bba675089f51a4af8eR36-R39

One could also define routes in the YAML files but our official devfile samples will have the route definition in the component[x].kubernetes.endpoints

jerolimov commented:
> @jerolimov Thanks for the detailed testing and reports. The routes need to be defined in the devfile this way https://github.com/devfile-samples/devfile-sample-code-with-quarkus/pull/7/files#diff-23d39e3ae288b080d8aa0995928cd3c432f887a75f0ab3bba675089f51a4af8eR36-R39
> 
> One could also define routes in the YAML files but our official devfile samples will have the route definition in the component[x].kubernetes.endpoints

Hi @maysunfaisal, I understand that it should be defined somewhere. But when its not defined the backend should not create broken resource definitions. I suggest that the `Route` is not exported in this case. Can you convert it to a pointer so that the JSON doesn't contain it anymore in this case?

divyanshiGupta commented:
> > @jerolimov Thanks for the detailed testing and reports. The routes need to be defined in the devfile this way https://github.com/devfile-samples/devfile-sample-code-with-quarkus/pull/7/files#diff-23d39e3ae288b080d8aa0995928cd3c432f887a75f0ab3bba675089f51a4af8eR36-R39
> > One could also define routes in the YAML files but our official devfile samples will have the route definition in the component[x].kubernetes.endpoints
> 
> Hi @maysunfaisal, I understand that it should be defined somewhere. But when its not defined the backend should not create broken resource definitions. I suggest that the `Route` is not exported in this case. Can you convert it to a pointer so that the JSON doesn't contain it anymore in this case?

@maysunfaisal I agree with @jerolimov here that backend should not return a broken resource definition since the add flows will break. In case the route is not defined let's not return it.

divyanshiGupta commented:
@maysunfaisal I tested the UI PR with the current backend PR and it seems to work fine expect for the issue mentioned above. Resource URI is getting replaced with inline resource yaml and the api call is also successful with the new payload. All the desired resources are also returned but the route resource is broken as mentioned above. 


https://user-images.githubusercontent.com/20724543/196163901-7499a168-7a35-4fe4-9ced-a139e584534c.mov

cc: @jerolimov @invincibleJai

maysunfaisal commented on pkg/devfile/resources.go:
```diff
@@ -0,0 +1,237 @@
+				break
+			}
+
+			src := parser.YamlSrc{
+				Data: []byte(component.Kubernetes.Inlined),
+			}
+			fs := afero.Afero{Fs: afero.NewOsFs()}
+			values, err := parser.ReadKubernetesYaml(src, fs)
```
Good suggestion; I will put in a update to the library api method params.

but for now, I might go with `NewMemMapFs()`, dont really want to do `YAML.Unmarshal` because that would be just duplicating the devfile/library code in here to read the contents.

maysunfaisal commented:
@divyanshiGupta @jerolimov I have rebased and addressed the PR concerns as well as our internal Slack thread conversation. PTAL.

I have tested 3 cases:
1. Importing devfile sample without inline, it falls back to the current master branch approach. Successful.
2. Importing devfile sample with inline including deployment, service and endpoint. Successful.
3. Importing devfile sample with inline with only deployment, where service and endpoint are missing. Errors Out:

<img width="1652" alt="Screen Shot 2022-10-18 at 2 02 04 PM" src="https://user-images.githubusercontent.com/31771087/196509714-1253a3b4-0843-4be1-a995-93d09f369518.png">

jerolimov commented on pkg/devfile/types.go:
```diff
@@ -8,31 +8,31 @@ import (
 	DeployResource appsv1.Deployment `json:"deployResource"`
-	Service        corev1.Service    `json:"service"`
-	Route          routev1.Route     `json:"route"`
+	Service        *corev1.Service   `json:"service"`
+	Route          *routev1.Route    `json:"route"`
```
This pointers return this json if the Service or route is not defined:

```json
{
  /* ... */
  "service": null,
  "route": null
}
```

You can omit this null values, this would fix the crash you mentioned here https://github.com/openshift/console/pull/12000#issuecomment-1282804735. But it still need some frontend changes of PR #12168. So its up to you if you want apply this small change or not:

```suggestion
	Service        *corev1.Service   `json:"service,omitempty"`
	Route          *routev1.Route    `json:"route,omitempty"`
```

maysunfaisal replied:
@jerolimov oops, I have committed this. if you can, please lgtm it again, thanks!

divyanshiGupta commented:
> I observed one thing we should check on Slack. The Deployment contains a "variable" "{{CONTAINER_IMAGE}}" as image container.

@maysunfaisal It seems that this should be replaced with the image-name in the devfile. As @jerolimov mentioned its fine for phase 1 but lets improve it lates and make sure all variables are replaced with their respective values.

divyanshiGupta commented:
/label docs-approved
/label px-approved

jerolimov commented:
/lgtm

divyanshiGupta commented:
/retest
